[jit] Use computed goto to dispatch to next instruction in interpreter.

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60211 [jit] Use computed goto to dispatch to next instruction in interpreter.**

Summary:
Refactor the interpreter loop a bit to use `goto` to jump to the next
instruction at the end of each instruction body, only on supported compiler.
In some microbenchmark this can reduce the branch misprediction rate of
interpreter loop by 10x. Altough there isn't much improvement for running
time in real cases, it's potentially still beneficial to do this optimization
and refactor the loop to be more readable in the meanwhile.

Test Plan:
CI

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29211283](https://our.internmc.facebook.com/intern/diff/D29211283)

zhxchen17 committed:
[wip][jit] Use computed loop to dispatch to next instruction in interpreter.

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29211283).

zhxchen17 committed:
Update on "[wip][jit] Use computed loop to dispatch to next instruction in interpreter."


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29211283](https://our.internmc.facebook.com/intern/diff/D29211283)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit] Use computed goto to dispatch to next instruction in interpreter."


Summary:
Refactor the interpreter loop a bit to use `goto` to jump to the next
instruction at the end of each instruction body, only on supported compiler.
In some microbenchmark this can reduce the branch misprediction rate of
interpreter loop by 10x. Altough there isn't much improvement for running
time in real cases, it's potentially still beneficial to do this optimization
and refactor the loop to be more readable in the meanwhile.

Test Plan:
CI

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29211283](https://our.internmc.facebook.com/intern/diff/D29211283)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29211283).

zhxchen17 committed:
Update on "[jit] Use computed goto to dispatch to next instruction in interpreter."


Summary:
Refactor the interpreter loop a bit to use `goto` to jump to the next
instruction at the end of each instruction body, only on supported compiler.
In some microbenchmark this can reduce the branch misprediction rate of
interpreter loop by 10x. Altough there isn't much improvement for running
time in real cases, it's potentially still beneficial to do this optimization
and refactor the loop to be more readable in the meanwhile.

Test Plan:
CI

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29211283](https://our.internmc.facebook.com/intern/diff/D29211283)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29211283).

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -208,6 +208,37 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
   }
 
+#if defined(__GNUC__) || defined(__clang__)
+#define USE_COMPUTED_GOTO
```
Nit:
Should we use a more specific name like JIT_USE_COMPUTED_GOTO and `undef` it at end of this logic block?

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -208,6 +208,37 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
+// Also blocks should be ended properly with either `INST_NEXT` (for going
+// to the next instruction), or `INST_DISPATCH` (for jumping to a computed
+// position using `INST_FETCH`).
+#define INST_FETCH(X) (frame.function->instructions_[frame.pc += (X)])
```
Optional:
Maybe code will be cleaner if change `INST_FETCH(X)` to:

`inst = (frame.function->instructions_[frame.pc += (X)])`

gmagogsfm reviewed (approved):
Looks good, only some optional comments

zhxchen17 committed:
Update on "[jit] Use computed goto to dispatch to next instruction in interpreter."


Summary:
Refactor the interpreter loop a bit to use `goto` to jump to the next
instruction at the end of each instruction body, only on supported compiler.
In some microbenchmark this can reduce the branch misprediction rate of
interpreter loop by 10x. Altough there isn't much improvement for running
time in real cases, it's potentially still beneficial to do this optimization
and refactor the loop to be more readable in the meanwhile.

Test Plan:
CI

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29211283](https://our.internmc.facebook.com/intern/diff/D29211283)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29211283).
